buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -714,20 +714,95 @@ public static ByteBuf readBytes(ByteBufAllocator alloc, ByteBuf buffer, int leng
         }
     }
 
-    static int lastIndexOf(AbstractByteBuf buffer, int fromIndex, int toIndex, byte value) {
+    static int lastIndexOf(final AbstractByteBuf buffer, int fromIndex, final int toIndex, final byte value) {
         assert fromIndex > toIndex;
         final int capacity = buffer.capacity();
         fromIndex = Math.min(fromIndex, capacity);
-        if (fromIndex < 0 || capacity == 0) {
+        if (fromIndex <= 0) { // fromIndex is the exclusive upper bound.
             return -1;
         }
-        buffer.checkIndex(toIndex, fromIndex - toIndex);
+        final int length = fromIndex - toIndex;
+        buffer.checkIndex(toIndex, length);
+        if (!PlatformDependent.isUnaligned()) {
+            return linearLastIndexOf(buffer, fromIndex, toIndex, value);
+        }
+        final int longCount = length >>> 3;
+        int offset = fromIndex;
+        if (longCount > 0) {
+            final ByteOrder nativeOrder = ByteOrder.nativeOrder();
+            final boolean isNative = nativeOrder == buffer.order();
+            final boolean useLE = nativeOrder == ByteOrder.LITTLE_ENDIAN;
+            final long pattern = SWARUtil.compilePattern(value);
+            for (int i = 0; i < longCount; i++) {
+                // use the faster available getLong
+                final long word = useLE? buffer._getLongLE(offset - Long.BYTES)
+                        : buffer._getLong(offset - Long.BYTES);
+                final long result = SWARUtil.applyPattern(word, pattern);
+                if (result != 0) {
+                    // used the oppoiste endianness since we are looking for the last index.
+                    return offset - 1 - SWARUtil.getIndex(result, !isNative);
+                }
+                offset -= Long.BYTES;
+            }
+        }
+        return unrolledLastIndexOf(buffer, offset, length & 7, value);
+    }
+
+    private static int linearLastIndexOf(final AbstractByteBuf buffer, final int fromIndex, final int toIndex,
+                                         final byte value) {
         for (int i = fromIndex - 1; i >= toIndex; i--) {
             if (buffer._getByte(i) == value) {
                 return i;
             }
         }
+        return -1;
+    }
 
+    private static int unrolledLastIndexOf(final AbstractByteBuf buffer, final int fromIndex, final int byteCount,
```

Comment:
Did you compare this `unrolledLastIndexOf` to calling `linearLastIndexOf` with adjusted range?

Reply:
Previous research has shown that manually unrolled loops improves performance for `size=7` benchmark case(https://github.com/netty/netty/pull/10737#issuecomment-720481198).

By the way, I will add an updated comparison.

Reply:
Manual unrolling results in better performance compared to a linear approach. (size > 1)

[benchmark](https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/jchrys/b4c0f164e8575c88e4c4665fe6532a89/raw/b5184b9df7e41218e5c14a69471f15c5291d71c6/lastIndexOfLinear.json,https://gist.githubusercontent.com/jchrys/b4c0f164e8575c88e4c4665fe6532a89/raw/b5184b9df7e41218e5c14a69471f15c5291d71c6/lastIndexOfUnroll.json)

[linear benchmark source code](https://github.com/jchrys/netty/blob/4.1-bytebuf-last-index-linear-bench/microbench/src/main/java/io/netty/microbench/buffer/ByteBufLastIndexOfBenchmark.java)
[manual unroll benchmark source code
](https://github.com/jchrys/netty/blob/4.1-bytebuf-last-index-unroll-bench/microbench/src/main/java/io/netty/microbench/buffer/ByteBufLastIndexOfBenchmark.java)